### PR TITLE
PC-1044 Streaming runs

### PR DESCRIPTION
**Pull request by paulcjh**
# Pull request outline

## Checklist:
- [ ] Docs updated
- [ ] Tests written
- [ ] Check for breaking changes in Resource
- [ ] Version bumped

Added:
- A list of new things

Changed:
- A list of things changed

Removed:
- A list of things removed

## Related issues:
_none_

___

- _You can add a reference to an issue using hash followed by the issue number_
- _If a checklist item is to be ignored it must be struck through_

### CODE REVIEW

File: pipeline/cloud/schemas/runs.py [hunk 1]
```diff
@@ -130,6 +131,11 @@ def from_object(cls, obj: t.Any):
             except (TypeError, OverflowError):
                 return cls.pkl
             return cls.array
+        # Also check if class name == Stream
+        elif hasattr(obj, "iterable") or (
+            hasattr(obj, "__name__") and obj.__name__ == "Stream"
+        ):
```
Comment (rossgray) [inline pipeline/cloud/schemas/runs.py:137]: I'm not too familiar with this code but can we not just check if `isinstance(obj, Stream)`?

File: pipeline/container/routes/v4/runs.py [hunk 2]
```diff
@@ -80,17 +85,96 @@ async def run(
...
+                                run_schemas.RunOutput(
+                                    type=output.type,
+                                    value=None,
+                                    file=output.file,
+                                )
+                            )
+                            streaming_outputs_to_move.append(output)
+
+                    for output in streaming_outputs_to_move:
+                        streaming_outputs.remove(output)
```
Comment (rossgray) [inline pipeline/container/routes/v4/runs.py:140]: I don't quite understand what's going on here. From what I can see, it looks like if we reach the end of our stream we add a `None` to the final outputs in between the static outputs and the next streaming outputs?

**Comment by rossgray** [comment]
@paulcjh I wonder if we should make the streaming endpoint in the container a separate endpoint, as we're doing with the backend API, since previously we've tried to keep `/v4/runs` consistent between the container and pcore?

File: pipeline/cloud/schemas/runs.py (hunk 1 above)
Comment (paulcjh) [reply pipeline/cloud/schemas/runs.py:137]: I think I did this because your suggestion would cause a circular dependency

**Comment by rossgray** [comment]
the only other thought I had (and maybe we can always do this later) was whether it makes sense to return the full run result schema every time, or just the raw outputs? Since I have a feeling all of the serialising JSON in and out of Pydantic objects may add a bit of overhead. We could always add a separate endpoint for that later though. Maybe also an endpoint where the results are streamed directly from the resource too without going via catalyst and pcore

File: pipeline/cloud/schemas/runs.py (hunk 1 above)
Comment (rossgray) [reply pipeline/cloud/schemas/runs.py:137]: ah ok. Moving the `Stream` class into a separate module may help

**Review by plutopulp — APPROVED** [review]
Nice to get this in! We've checked over this a fair few times but had a last look and LGTM

**Pull request merged**